app/controllers/supplementary/DocumentsProducedController.scala

```diff
@@ -34,29 +41,130 @@ class DocumentsProducedController @Inject()(
-    customsCacheService.fetchAndGetEntry[DocumentsProduced](supplementaryCacheId, formId).map {
-      case Some(data) => Ok(documents_produced(appConfig, form.fill(data)))
-      case _          => Ok(documents_produced(appConfig, form))
-    }
+    customsCacheService
+      .fetchAndGetEntry[DocumentsProducedData](supplementaryCacheId, formId)
+      .map {
+        case Some(data) => Ok(documents_produced(appConfig, form, data.documents))
+        case _          => Ok(documents_produced(appConfig, form, Seq()))
+      }
   }
 
   def saveForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[DocumentsProduced]) =>
-          Future.successful(BadRequest(documents_produced(appConfig, formWithErrors))),
-        form =>
-          customsCacheService.cache[DocumentsProduced](supplementaryCacheId, formId, form).map { _ =>
-            Redirect(controllers.supplementary.routes.SummaryPageController.displayPage())
+    val boundForm = form.bindFromRequest()
+    val actionTypeOpt = request.body.asFormUrlEncoded.flatMap(FormAction.fromUrlEncoded(_))
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[DocumentsProducedData](supplementaryCacheId, formId)
+      .map(_.getOrElse(DocumentsProducedData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[DocumentsProduced]) =>
+            Future.successful(BadRequest(documents_produced(appConfig, formWithErrors, cache.documents))),
+          validForm =>
+            actionTypeOpt match {
+              case Some(Add)             => addItem(validForm, cache)
+              case Some(SaveAndContinue) => saveAndContinue(validForm, cache)
+              case Some(Remove(values))  => removeItem(retrieveItem(values), cache)
+              case _                     => errorHandler.displayErrorPage()
+          }
+        )
+    }
+  }
+
+  private def saveAndContinue(
+    userInput: DocumentsProduced,
+    cachedData: DocumentsProducedData
+  )(implicit request: AuthenticatedRequest[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.documents) match {
+      case (document, Seq())     => handleSaveAndContinueEmptyCache(document)
+      case (document, documents) => handleSaveAndContinueCache(document, documents)
+    }
+
+  private def handleSaveAndContinueCache(document: DocumentsProduced, documents: Seq[DocumentsProduced])(
+    implicit request: AuthenticatedRequest[_]
+  ) =
+    document match {
+      case _ if documents.length >= maxNumberOfItems =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.maximumAmount.error")), document, documents)
+
+      case _ if documents.contains(document) =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.duplicated")), document, documents)
+
+      case _ if document.isDefined =>
```

Comment:
This and next case are very similar, my propose is to use

`val updatedCache = if(document.isDefined) DocumentsProducedData(documents :+ document) else DocumentsProducedData(documents)`

To avoid having to cases, in both of them we will make a redirect, in one case it'll be one unnecessary call to the database. You can improve my propose and based on this cache or don't cache data.

Reply:
I think we should reuse the code there. 
In the case the `! document.isDefined`,AKA _(not defined)_, we do not need to do a DB call.

Reply:
yeah, so I think that we should use solution to not do db call when you don't need it

Reply:
```
if (document.isDefined)
      customsCacheService
        .cache[DocumentsProducedData](supplementaryCacheId, formId, DocumentsProducedData(documents :+ document))
    Future.successful(Redirect(controllers.supplementary.routes.SummaryPageController.displayPage()))
```

This is my suggestion, 
One of my golden rules is to avoid `else` keyword. So the common code always gets run and the exception is caught by the if statement. If also read better :)
What do you think?

Reply:
Yep, it's really good approach, but here when you have call to database we should do like this, so probably the easiest one will be with stay with cases, because now the call to database will be started, but you're not sure that it'll finish with successful before redirection